File: src/main/java/com/procurement/point/controller/PublicTenderController.java
package com.procurement.point.controller;

import com.procurement.point.model.dto.offset.OffsetDto;
import com.procurement.point.model.dto.record.RecordPackageDto;
import com.procurement.point.model.dto.release.ReleasePackageDto;
import com.procurement.point.service.PublicTenderService;
import java.time.LocalDateTime;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@CrossOrigin(maxAge = 3600)
@RestController
@RequestMapping(value = "/tender")
public class PublicTenderController {

    private final PublicTenderService publicService;

    public PublicTenderController(final PublicTenderService publicService) {
        this.publicService = publicService;
    }

    @RequestMapping(method = RequestMethod.GET, value = "/{cpid}")
    public ResponseEntity<RecordPackageDto> getRecordPackage(@PathVariable(value = "cpid") final String cpid,
                                                             @RequestParam(value = "offset", required = false)
                                                             @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
                                                             final LocalDateTime
                                                                 offset) {
        return new ResponseEntity<>(publicService.getRecordPackage(cpid, offset), HttpStatus.OK);
    }

    @RequestMapping(method = RequestMethod.GET, value = "/{cpid}/{ocid}")
    public ResponseEntity<ReleasePackageDto> getReleasePackage(@PathVariable(value = "cpid") final String cpid,
                                                               @PathVariable(value = "ocid") final String ocid,
                                                               @RequestParam(value = "offset", required = false)
                                                               @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
                                                                   final LocalDateTime
                                                                   offset) {
        return new ResponseEntity<>(publicService.getReleasePackage(cpid, ocid, offset), HttpStatus.OK);
    }

    @RequestMapping(method = RequestMethod.GET)
    public ResponseEntity<OffsetDto> getByOffset(@RequestParam(value = "offset")
                                                 @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) final
                                                 LocalDateTime offset,
                                                 @RequestParam(value = "limit") final Integer limit) {
        return new ResponseEntity<>(publicService.getByOffset(offset, limit), HttpStatus.OK);
    }
}

File: src/main/kotlin/com/procurement/point/controller/PublicTenderController.kt
package com.procurement.point.controller

import com.procurement.point.exception.ParamException
import com.procurement.point.model.dto.OffsetDto
import com.procurement.point.model.dto.RecordPackageDto
import com.procurement.point.model.dto.ReleasePackageDto
import com.procurement.point.service.PublicTenderService
import org.springframework.format.annotation.DateTimeFormat
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.time.LocalDateTime
import javax.servlet.http.HttpServletRequest

@RestController
@CrossOrigin(maxAge = 3600)
@RequestMapping("/tenders")
class PublicTenderController(private val publicService: PublicTenderService) {

    @GetMapping
    fun getByOffset(@RequestParam(value = "offset", required = false)
                    @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
                    offset: LocalDateTime?,
                    @RequestParam(value = "limit", required = false) limitParam: Int?,
                    request: HttpServletRequest): ResponseEntity<OffsetDto> {
        checkParams(request.parameterNames.toList(), listOf("offset", "limit"))
        return ResponseEntity(publicService.getByOffset(offset, limitParam), HttpStatus.OK)
    }

    @GetMapping("/{cpid}")
    fun getRecordPackage(@PathVariable(value = "cpid") cpid: String,
                         @RequestParam(value = "offset", required = false)
                         @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
                         offset: LocalDateTime?,
                         request: HttpServletRequest): ResponseEntity<RecordPackageDto> {
        checkParams(request.parameterNames.toList(), listOf("offset"))
        return ResponseEntity(publicService.getRecordPackage(cpid, offset), HttpStatus.OK)
    }

    @GetMapping("/{cpid}/{ocid}")
    fun getReleasePackage(@PathVariable(value = "cpid") cpid: String,
                          @PathVariable(value = "ocid") ocid: String,
                          @RequestParam(value = "offset", required = false)
                          @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
                          offset: LocalDateTime?,
                          request: HttpServletRequest): ResponseEntity<ReleasePackageDto> {
        checkParams(request.parameterNames.toList(), listOf("offset"))
        return ResponseEntity(publicService.getRecord(cpid, ocid, offset), HttpStatus.OK)
    }

    @GetMapping("/cn")
    fun getByOffsetCn(@RequestParam(value = "offset", required = false)
                      @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
                      offset: LocalDateTime?,
                      @RequestParam(value = "limit", required = false) limitParam: Int?,
                      request: HttpServletRequest): ResponseEntity<OffsetDto> {
        checkParams(request.parameterNames.toList(), listOf("offset", "limit"))
        return ResponseEntity(publicService.getByOffsetCn(offset, limitParam), HttpStatus.OK)
    }

    @GetMapping("/plan")
    fun getByOffsetPlan(@RequestParam(value = "offset", required = false)
                        @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
                        offset: LocalDateTime?,
                        @RequestParam(value = "limit", required = false) limitParam: Int?,
                        request: HttpServletRequest): ResponseEntity<OffsetDto> {
        checkParams(request.parameterNames.toList(), listOf("offset", "limit"))
        return ResponseEntity(publicService.getByOffsetPlan(offset, limitParam), HttpStatus.OK)
    }

    private fun checkParams(parameterNames: List<String>, prams: List<String>) {
        parameterNames.forEach { paramRq ->
            if (!prams.contains(paramRq)) throw ParamException("Invalid parameter:'$paramRq'. Valid params: $prams")
        }
    }
}

